onnxruntime/core/graph/graph.cc

```diff
@@ -987,15 +987,22 @@ Status Graph::BuildConnections(std::vector<std::string>& outer_scope_node_args_c
           continue;
         }
 
-        auto output_arg_iter = resolve_context_.output_args.find(input_arg->Name());
+        const auto& input_arg_name = input_arg->Name();
+        auto output_arg_iter = resolve_context_.output_args.find(input_arg_name);
         if (resolve_context_.output_args.end() == output_arg_iter) {
           // No such output_arg matching this input_arg.
           // This input arg should be fed when running evaluation.
+
           // See if it's present in the outer scope. If so it will be 'fed' by the execution frame
           // providing access to the OrtValue from the outer scope. Pass the name back up so nodes can
           // be linked correctly at that level.
-          if (outer_scope_node_args.find(input_arg->Name()) != outer_scope_node_args.cend()) {
-            outer_scope_node_args_consumed.push_back(input_arg->Name());
+          if (outer_scope_node_args.find(input_arg_name) != outer_scope_node_args.cend()) {
+            // if it matches an outer scope value make sure it's not also a graph input or initializer
+            // as the ONNX checker currently allows a graph input to shadow an outer scope value
+            if (resolve_context_.inputs_and_initializers.find(input_arg_name) ==
```

Comment:
Minor nit: Conceptually, this belongs together with the search in line 991 (search the locally-defined names, which is the union of output_args and inputs and initializers).  While it has the same effect, it might be a bit easier to understand if we move this check up there (basically swapping the two checks in 999 and 1002 around).